### ISSUE TITLE
Building the AST is not practical because it runs tests - how to prevent that?

### ISSUE THREAD

**Issue by Andre0991** [clojure-emacs/clj-refactor.el#426]
`cljr-find-usages` and friends have to create the AST for the whole project.

This is not practical for my use case, as it would need to run all the tests, which might take a long time. Also, some tests actually break the AST creation with `Method code too large!`.

Is there are way to configure `clj-refactor` so that it ignore certain namespaces when building the AST?

**Comment by expez** [comment]
The easiest way to solve this would be wrap your tests in a function, which your test-runner calls, instead of having `(run-tests!)` as a top-level form.  That way your tests won't run as part of the analysis step performed by tools.analyzer.

There's code in place to ignore certain paths that's already been written on the middleware side of things.  It would be easy to expose that to the client as a config setting, and we should probably do that.

However, if you ignore your test directory, then renamings using cljr-rename-symbol won't be propagated to these namespaces because they haven't been analyzed!  This could be a source of both breakage and confusion :(

In other words, you'd unfortunately have to refactor a little even with a new config setting, e.g. to have a file with nothing except a require form and `(run-tests!)` as your the form.

**Comment by expez** [comment]
You should be able to apply this minimal patch manually, if you're so inclined, to see if it solves your problem @Andre0991: https://github.com/clojure-emacs/clj-refactor.el/pull/427

**Comment by Andre0991** [comment]
> The easiest way to solve this would be wrap your tests in a function, which your test-runner calls, instead of having (run-tests!) as a top-level form. That way your tests won't run as part of the analysis step performed by tools.analyzer.

The issue is that I use Midje, which runs the test when the test form is evaluated. I actually like this behaviour, it makes testing more interactive - but it does make evaluating the whole project impractical.

> However, if you ignore your test directory, then renamings using cljr-rename-symbol won't be propagated to these namespaces because they haven't been analyzed! This could be a source of both breakage and confusion :(
> In other words, you'd unfortunately have to refactor a little even with a new config setting, e.g. to have a file with nothing except a require form and (run-tests!) as your the form.

You're right, that would be confusing. I'm not sure if there's a good solution for this using Midje :(

Thanks for the patch. I manually applied it, but I got the following error:

```
error in process filter: cljr--get-error-value: Error in nrepl-refactor: java.util.concurrent.ExecutionException: java.lang.ClassCastException: java.lang.String cannot be cast to java.util.regex.Pattern
 at java.util.concurrent.FutureTask.report (FutureTask.java:122)
    java.util.concurrent.FutureTask.get (FutureTask.java:192)
    clojure.core$deref_future.invokeStatic (core.clj:2292)
    clojure.core$future_call$reify__8097.deref (core.clj:6894)
    clojure.core$deref.invokeStatic (core.clj:2312)
    clojure.core$deref.invoke (core.clj:2298)
    refactor_nrepl.find.find_symbol$find_symbol.invokeStatic (find_symbol.clj:242)
    refactor_nrepl.find.find_symbol$find_symbol.invoke (find_symbol.clj:235)
    clojure.lang.Var.invoke (Var.java:381)
    refactor_nrepl.middleware$find_symbol_reply.invokeStatic (middleware.clj:88)
    refactor_nrepl.middleware$find_symbol_reply.invoke (middleware.clj:85)
    refactor_nrepl.middleware$wrap_refactor$fn__22106.invoke (middleware.clj:203)
    
    ....
```

**Comment by expez** [comment]
Thanks, think we're missing a coercion step in the middleware to turn these into regexes.

**Comment by expez** [comment]
The new snapshot is out @Andre0991.  Let please me know if this is solved now.

**Comment by Andre0991** [comment]
Thanks @expez .
I haven't been able to make it ignore my tests files - is my regex wrong or something?

> Caused by: java.lang.IllegalStateException: refactor-nrepl is unable to build an AST for purgatory.logic.agreement.validation-test. tools.analyzer encountered the following problem: java.lang.RuntimeException: Method code too large!, compiling:(file:/Users/user/dev/project/test/project/logic/agreement/validation_test.clj:245:3)

This is my attempt:
```
(setq cljr-middleware-ignored-paths '("test\\.*"))
```

**Comment by expez** [comment]
@Andre0991  change the `\\` to `\` in either of those and you should be good to go (java regex)

**Comment by Andre0991** [comment]
@expez, for some reason it's always triggering the same error. How can I debug that? 🤔

It complains about `file:/Users/myUser/dev/nu/purgatory/test/purgatory/logic/agreement/validation_test.clj:245:3`.

My clj-refactor version is 20180826.2149.

I basically set `cljr-middleware-ignored-paths` in my dotfile as above (I removed the slash and tried to use other formats too).

**Comment by Andre0991** [comment]
Oh, that makes sense. 

Ignoring all paths with `/test/` would be OK for my current use case too.
But still, it won't work with `"test"` either :/

I'm suspecting nrepl is not actually considering it.

Let me know if you need more information.
